File: FanficlyTests/StubAO3Client.swift
import Foundation
@testable import Fanficly

/// Test double whose autocomplete responses are scripted, so resolution
/// logic (e.g. the relationship character-fallback) can be exercised.
final class StubAO3Client: AO3ClientProtocol, @unchecked Sendable {
    /// term (lowercased) → matches
    var autocompleteResponses: [String: [String]] = [:]
    private(set) var autocompleteCalls: [String] = []

    func autocomplete(field: AO3AutocompleteField, term: String) async throws -> [String] {
        autocompleteCalls.append(term)
        return autocompleteResponses[term.lowercased()] ?? []
    }

    // Unused protocol stubs.
    func login(username: String, password: String) async throws {}
    func logout() async {}
    func currentUsername() async -> String? { nil }
    func search(filters: AO3SearchFilters, page: Int) async throws -> AO3SearchResults {
        AO3SearchResults(works: [], totalPages: 0, currentPage: page)
    }
    func fetchAuthorWorks(username: String, page: Int) async throws -> AO3SearchResults {
        AO3SearchResults(works: [], totalPages: 0, currentPage: page)
    }
    func fetchBookmarks(username: String, page: Int) async throws -> AO3SearchResults {
        AO3SearchResults(works: [], totalPages: 0, currentPage: page)
    }
    func fetchWork(id: Int) async throws -> AO3WorkPayload {
        AO3WorkPayload(summary: .stub(id: id, title: "x"), chapters: [])
    }
    func fetchWorkMetadata(id: Int) async throws -> AO3WorkMetadata {
        AO3WorkMetadata(id: id, chapterCount: 1, totalChapters: 1, updatedAt: nil)
    }
    func fetchSubscriptions(username: String) async throws -> [AO3Subscription] { [] }
    func fetchComments(workId: Int, chapterId: Int?) async throws -> [AO3Comment] { [] }
    func postComment(workId: Int, chapterId: Int?, text: String) async throws {}
    func fetchFandomsInCategory(categoryName: String) async throws -> [BrowseFandom] { [] }
    func fetchPopularSnapshot() async throws -> PopularSnapshot {
        PopularSnapshot(fandoms: [], ships: [], characters: [])
    }
    func downloadEPUB(workId: Int) async throws -> URL {
        URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("\(workId).epub")
    }
    func exportWork(workId: Int, format: WorkExportFormat, filename: String) async throws -> URL {
        URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("\(filename).\(format.ext)")
    }
    func subscribeToWork(workId: Int) async throws {}
}

File: android/app/src/test/java/io/github/yennster/fanficly/StubAO3Client.kt
package io.github.yennster.fanficly

import io.github.yennster.fanficly.model.AO3SearchFilters
import io.github.yennster.fanficly.model.AutocompleteField
import io.github.yennster.fanficly.model.ExportFormat
import io.github.yennster.fanficly.model.SearchResults
import io.github.yennster.fanficly.model.WorkMetadata
import io.github.yennster.fanficly.model.WorkPayload
import io.github.yennster.fanficly.net.AO3Client
import java.io.File

/**
 * Scriptable [AO3Client] test double — the Kotlin port of the iOS
 * `StubAO3Client`. Only [autocomplete] is meaningful: it returns whatever the
 * [autocompleteResponses] map holds for the (case-insensitive) term, defaulting
 * to an empty list. Everything else throws so a test that accidentally hits the
 * network fails loudly.
 */
class StubAO3Client : AO3Client {
    var autocompleteResponses: Map<String, List<String>> = emptyMap()

    override suspend fun autocomplete(field: AutocompleteField, term: String): List<String> =
        autocompleteResponses[term.lowercase()] ?: emptyList()

    override suspend fun login(username: String, password: String) = throw NotImplementedError()
    override fun logout() = throw NotImplementedError()
    override fun currentUsername(): String? = null
    override fun isLoggedIn(): Boolean = false
    override suspend fun search(filters: AO3SearchFilters, page: Int): SearchResults = throw NotImplementedError()
    override suspend fun fetchAuthorWorks(username: String, page: Int): SearchResults = throw NotImplementedError()
    override suspend fun fetchBookmarks(username: String, page: Int): SearchResults = throw NotImplementedError()
    override suspend fun fetchWork(id: Int): WorkPayload = throw NotImplementedError()
    override suspend fun fetchWorkMetadata(id: Int): WorkMetadata = throw NotImplementedError()
    override suspend fun fetchFandomsInCategory(categoryName: String): List<io.github.yennster.fanficly.model.BrowseFandom> = throw NotImplementedError()
    override suspend fun fetchPopularSnapshot(): io.github.yennster.fanficly.model.PopularSnapshot = throw NotImplementedError()
    override suspend fun fetchComments(workId: Int, chapterId: Int?): List<io.github.yennster.fanficly.model.AO3Comment> = throw NotImplementedError()
    override suspend fun postComment(workId: Int, chapterId: Int?, text: String) = throw NotImplementedError()
    override suspend fun downloadEpub(workId: Int, cacheDir: File): File = throw NotImplementedError()
    override suspend fun exportWork(workId: Int, format: ExportFormat, filename: String, cacheDir: File): File = throw NotImplementedError()
    override suspend fun subscribeToWork(workId: Int) = throw NotImplementedError()
}
